R/tar-stars.R

```diff
@@ -0,0 +1,220 @@
+                            garbage_collection = targets::tar_option_get("garbage_collection"),
+                            deployment = targets::tar_option_get("deployment"),
+                            priority = targets::tar_option_get("priority"),
+                            resources = targets::tar_option_get("resources"),
+                            storage = targets::tar_option_get("storage"),
+                            retrieval = targets::tar_option_get("retrieval"),
+                            cue = targets::tar_option_get("cue")) {
+
+    rlang::check_installed("stars")
+
+    name <- targets::tar_deparse_language(substitute(name))
+
+    envir <- targets::tar_option_get("envir")
+
+    command <- targets::tar_tidy_eval(
+        expr = as.expression(substitute(command)),
+        envir = envir,
+        tidy_eval = tidy_eval
+    )
+
+    pattern <- targets::tar_tidy_eval(
+        expr = as.expression(substitute(pattern)),
+        envir = envir,
+        tidy_eval = tidy_eval
+    )
+
+    targets::tar_target_raw(
+        name = name,
+        command = command,
+        pattern = pattern,
+        packages = packages,
+        library = library,
+        format = create_format_stars(
+            driver = driver,
+            options = options,
+            proxy = TRUE,
+            mdim = mdim,
+            ncdf = ncdf,
+            ...
+        ),
+        repository = repository,
+        iteration = iteration,
+        error = error,
+        memory = memory,
+        garbage_collection = garbage_collection,
+        deployment = deployment,
+        priority = priority,
+        resources = resources,
+        storage = storage,
+        retrieval = retrieval,
+        cue = cue
+    )
+}
+
+create_format_stars <- function(driver, options, proxy, mdim, ncdf, ...) {
+
+    driver <- driver %||% "GTiff"
+    options <- options %||% character(0)
+
+    # get list of drivers available for writing depending on what the user's GDAL supports
+    drv <- sf::st_drivers(what = "raster")
+    drv <- drv[drv$write, ]
+
+    driver <- rlang::arg_match0(driver, drv$name)
+
+    READ_FUN <- "stars::read_stars"
+    WRITE_FUN <- "stars::write_stars"
+
+    if (mdim) {
+        READ_FUN <- "stars::read_mdim"
+        WRITE_FUN <- "stars::write_mdim"
+    }
+
+    if (ncdf && requireNamespace("ncmeta")) {
+        READ_FUN <- "stars::read_ncdf"
+    }
+
+    .read_stars <- eval(substitute(function(path) {
+        FUN(path, proxy = proxy)
+    }, list(FUN = str2lang(READ_FUN),
+            proxy = proxy)))
+
+    # TODO: should multidimensional array use the same `options` as 2D?
+    .write_stars <- eval(substitute(function(object, path) {
+        FUN(
+            object,
+            path,
+            driver = driver,
+            overwrite = TRUE,
+            options = options
+        )
+    }, list(FUN = str2lang(WRITE_FUN),
+            driver = driver,
+            options = options)))
```

Comment:
Rather than `eval(substitute(...` we now have the option to pass env vars to the custom format using `targets::tar_resources_custom_format()`.  See the updates to `tar_terra_rast()`, e.g. [here](https://github.com/njtierney/geotargets/blob/e1b039f372f5f82c9cd08ac27c319c08955f0c75/R/tar-terra-rast.R#L85-L91) and [here](https://github.com/njtierney/geotargets/blob/e1b039f372f5f82c9cd08ac27c319c08955f0c75/R/tar-terra-rast.R#L104-L108).  I think we should stick to this new pattern as it's a bit easier to comprehend, IMO.

Reply:
> Rather than `eval(substitute(...` we now have the option to pass env vars to the custom format using `targets::tar_resources_custom_format()`. 

I have this mostly implemented, it does make the code look cleaner. 

I'm running into some issues with arguments not being properly passed when using `tar_resources_custom_format()` though. I tinkered with it a bit last night and will take a fresh look tonight. If I can't figure it out I will push the changes anyway to see if you can spot what I am doing wrong.

Reply:
It appears the source of my problem is that environment variables that are created via `tar_resources_custom_format()` are not available / not working properly in the `tar_format()` `read` method. 

I managed to localize the problems I was having to that, so for time being continuing to use `eval(substitute(...))` for read method dynamics. Using the env vars for write method, though. 

I have not yet had a chance to look into whether this is "expected" or if it is an issue in targets.

All tests are passing with expected values for the current code. If code to access env vars in read method is uncommented, the `tar_stars_proxy()` tests fail because default `proxy=FALSE` is used because the env var GEOTARGETS_PROXY is apparently unset.

Reply:
I see that you've commented this well, so I'm going to go ahead and merge this PR.  I think once we have all had some more practice with metaprogramming with targets, we'll go through and harmonize things across functions a bit.